### PR TITLE
Update polling and timeout message handling | CTT-188

### PR DESCRIPTION
**Pull request by codingLogan**
This PR implements some edge-case fixes that have been reported by customers, which cause the user to get stuck on the Connecting screen, or get the wrong screen.

1. Sometimes the first response from the backend APIs already signal that the job is done, and in an error state.  The widget incorrectly would keep on polling waiting for a _different response_ from the first one.  _This was recreated by using a Slow 3G network speed emulation, and using verification and No DDA member_.
2. Not aggregating fix - we have a handler for when we don't have an explicit actionable error code, but we are in a connection_status that is an error.  Updates were made to allow that state to be reached.
3. The two above fixes also exposed another problem.  The Connecting screen is supposed to send a timeout post message after 60 seconds.  Polling timing has changed in the last half-year, and this the timeout message was out of sync.  This problem has also been addressed because the above two fixes changed that code as well.

Side Note - While testing the above, it was also discovered that sometimes the widget would show a false-positive success screen in No DDA scenarios.  **The fix for this is being investigated as a backend change**.

Fixes
* 188
* 190
* 191

## Test Results

✅ Vitest
✅ New Playwright OAuth tests
✅ Cypress Tests (when npm linked to these changes)

### CODE REVIEW

File: src/hooks/usePollMember.tsx [hunk 1]
```diff
@@ -31,6 +33,7 @@ export function usePollMember() {
 
   const { optOutOfEarlyUserRelease, memberPollingMilliseconds, useWebSockets } =
     useSelector(getExperimentalFeatures)
+  const mode = useSelector((state: RootState) => state.config?.mode ?? AGG_MODE)
```
Comment (wesrisenmay-mx) [inline src/hooks/usePollMember.tsx]: It looks like mode is already defaulted in handlePollingResponse, so I don't think you need to default it here

File: src/hooks/__tests__/usePollMember-test.tsx [hunk 2]
```diff
@@ -871,4 +876,137 @@ describe('usePollMember', () => {
...
+    })
+
+    await waitFor(
+      () => {
+        expect(states.some((s) => s.pollingIsDone === true)).toBe(true)
+      },
+      { timeout: 1000 },
+    )
+
+    expect(states.some((s) => s.pollingIsDone === true)).toBe(true)
```
Comment (mwclemy) [inline src/hooks/__tests__/usePollMember-test.tsx]: Is this second assertion necessary? It looks like the previous assertion already waits longer than the `100ms memberPollingMilliseconds` interval.

File: src/views/connecting/Connecting.js [hunk 3]
```diff
@@ -93,12 +96,12 @@ export const Connecting = (props) => {
...
 
   function handleMemberPoll(pollingState) {
-    // If we have been polling for more than 15 attempts (60 seconds)
-    // show the timeout message
+    // If polling has run longer than the timeout threshold, show timeout.
     // Unless this is a PENDING member, then we don't show the timeout
     // since PENDING may take much longer to resolve.
     if (
-      pollingState.pollingCount > 15 &&
+      pollingStartedAtRef.current !== null &&
```
Comment (mwclemy) [inline src/views/connecting/Connecting.js:103]: Can we add a test for this timeout?

File: src/utilities/__tests__/pollers-test.js [hunk 4]
```diff
@@ -60,11 +65,79 @@ describe('handlePollingResponse', () => {
...
+      },
+    }
+
+    const [stopPolling, message] = handlePollingResponse(pollingState, VERIFY_MODE)
+
+    expect(stopPolling).toEqual(false)
+    expect(message).toEqual(CONNECTING_MESSAGES.VERIFYING)
+  })
+
+  test('should stop polling on second consecutive CONNECTED update in verify mode', () => {
```
Comment (mwclemy) [inline src/utilities/__tests__/pollers-test.js]: Just for my own learning, why does verify mode need two `CONNECTED` updates?

File: src/utilities/__tests__/pollers-test.js [hunk 5]
```diff
@@ -161,7 +233,30 @@ describe('handlePollingResponse', () => {
...
+
+          expect(message).toEqual(CONNECTING_MESSAGES.ERROR)
+          expect(stopPolling).toEqual(true)
+        }
+      })
+    })
+
+    it('should stop polling when a terminal error code is present, oauth is true, and there is no previous response', () => {
+      ErrorStatuses.forEach((status) => {
+        if (isNotChallenged(status)) {
```
Comment (wesrisenmay-mx) [inline src/utilities/__tests__/pollers-test.js]: Any time I see an if inside a test it gives me pause. Instead of checking if the status is challenged inside each of these tests, why don't we just filter out the challenged status from this list? Then we don't have to put this if statement everywhere.

const nonChallengedErrorStatuses = ErrorStatuses.filter



...


nonChallengedErrorStatuses.forEach

File: src/utilities/__tests__/pollers-test.js (hunk 4 above)
Comment (wesrisenmay-mx) [reply src/utilities/__tests__/pollers-test.js]: I also wonder this

File: README.md [hunk 6]
```diff
@@ -67,6 +67,38 @@ You need to pass an object containing API endpoint callbacks as the `apiValue` p
 8. Update translations and documentation as needed
 9. Open Pull Request
 
+## Troubleshooting Polling and Websockets
```
Comment (wesrisenmay-mx) [inline README.md]: Should we just add these console.logs, but put a flag somewhere that turns them on or off?

File: src/views/connecting/Connecting.js (hunk 3 above)
Comment (codingLogan) [reply src/views/connecting/Connecting.js:103]: Yes, there's a significant portion added to the Connecting-test file for this timeout testing.  https://github.com/mxenabled/connect-widget/pull/335/changes#diff-db119e4b30337503bcd92b0a4fed5287911abf3f57f9ec259570aad2f34eabd0R67

File: src/utilities/__tests__/pollers-test.js (hunk 4 above)
Comment (codingLogan) [reply src/utilities/__tests__/pollers-test.js]: This particular scenario is one reason I would like reviews from some others because it is odd.  I'm open to suggestions of handling this differently.  I'll find where I placed my comments about this and add them to the test file.

The gist of the problem, is that an asynchronous edge case exists...  What happens behind the scenes of a member record that hits the No DDA flow, apparently is this.
1. The member goes to CONNECTED (no errors) - yay success!
2. The member then goes to IMPEDED (with errors) - oh no, error...

In some rare instances, 1 in 6, or sometimes 1 in 20 attempts, the widget would catch the CONNECTED status and show the success screen.  But, it should actually be showing the IMPAIRED screen...  With a second confirmed CONNECTED message we're more confident that it should actually be a success instead of an error.

Comment (wesrisenmay-mx) [reply src/utilities/__tests__/pollers-test.js]: This sounds like it's a backend problem, no?

File: README.md (hunk 6 above)
Comment (codingLogan) [reply README.md]: I debated on that.  The "flag somewhere" is the part that I didn't want to add to the already large PR 😂 .

My vote is that if we find ourselves regularly debugging this area, we can add something in that is more permanent.  I at least wanted to get the strategy that I used documented.

File: src/utilities/__tests__/pollers-test.js (hunk 4 above)
Comment (codingLogan) [reply src/utilities/__tests__/pollers-test.js]: Potentially?  If there is a bug, it would be with the `is_being_aggregated` flag.  That should likely stay `true` until completely after the jobs are done AND NO DDA is checked for verification involved jobs.

I'll raise some questions about this behavior, but for now this is a workaround we can handle.

Comment (mwclemy) [reply src/utilities/__tests__/pollers-test.js]: This sounds like a backend issue to me as well. I'm also wondering whether it's expected for a member with no DDA to be marked as `CONNECTED` and then `IMPEDED`, or if it should go straight to `IMPEDED`.

Comment (wesrisenmay-mx) [reply src/utilities/__tests__/pollers-test.js]: This feels like a monkey patch that is adding tech debt and complexity to our widget, when the problem really should be solved on the backend. Is adding this worth it? Do we have a ticket to remove this as soon as it's fixed on the backend? Has the backend prioritized fixing this, or is this monkey patch going to delay the backend getting the proper fix in?

Comment (codingLogan) [reply src/utilities/__tests__/pollers-test.js]: I'm going to leave the code in place while I personally investigate the backend behaviors.  Depending on what is found we can decide to revert the monkey-patch behavior, or adjust as needed to correctly detect no dda scenarios with both polling and/or websockets.

Best scenario, is that we can figure out the complete picture back-to-front, and then move this entire set of changes forward after we apply whatever fix is necessary.

**Review by wesrisenmay-mx — DISMISSED** [review]

**Review by wesrisenmay-mx — APPROVED** [review]
